vault/logical_system.go

```diff
@@ -1406,6 +1456,210 @@ func (b *SystemBackend) handleRotate(
+		resp.Data["creation_time"] = cubbyResp.Data["creation_time"]
+	}
+
+	return resp, nil
+}
+
+func (b *SystemBackend) handleWrappingRewrap(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	// If a third party is unwrapping (rather than the calling token being the
+	// wrapping token) we detect this so that we can revoke the original
+	// wrapping token after reading it. Right now wrapped tokens can't unwrap
+	// themselves, but in case we change it, this will be ready to do the right
+	// thing.
+	var thirdParty bool
+
+	token := data.Get("token").(string)
+	if token != "" {
+		thirdParty = true
+	} else {
+		token = req.ClientToken
+	}
+
+	if thirdParty {
+		defer b.Core.tokenStore.Revoke(token)
+	}
+
+	// Fetch the original TTL
+	cubbyReq := &logical.Request{
+		Operation:   logical.ReadOperation,
+		Path:        "cubbyhole/wrapinfo",
+		ClientToken: token,
+	}
+	cubbyResp, err := b.Core.router.Route(cubbyReq)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("error looking up wrapping information: %v", err)), nil
+	}
+	if cubbyResp == nil {
+		return logical.ErrorResponse("no information found; wrapping token may be from a previous Vault version"), nil
+	}
+	if cubbyResp != nil && cubbyResp.IsError() {
+		return cubbyResp, nil
+	}
+	if cubbyResp.Data == nil {
+		return logical.ErrorResponse("wrapping information was nil; wrapping token may be from a previous Vault version"), nil
+	}
+
+	// Set the creation TTL on the request
+	creationTTLRaw := cubbyResp.Data["creation_ttl"]
+	if creationTTLRaw == nil {
+		return logical.ErrorResponse("creation_ttl value in wrapping information was nil"), nil
+	}
+	creationTTL, err := cubbyResp.Data["creation_ttl"].(json.Number).Int64()
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("error reading creation_ttl value from wrapping information: %v", err)), ErrInternalError
+	}
+
+	// Fetch the original response and return it as the data for the new response
+	cubbyReq = &logical.Request{
+		Operation:   logical.ReadOperation,
+		Path:        "cubbyhole/response",
+		ClientToken: token,
+	}
+	cubbyResp, err = b.Core.router.Route(cubbyReq)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("error looking up wrapping information: %v", err)), nil
+	}
+	if cubbyResp == nil {
+		return logical.ErrorResponse("no information found; wrapping token may be from a previous Vault version"), nil
+	}
+	if cubbyResp != nil && cubbyResp.IsError() {
+		return cubbyResp, nil
+	}
+	if cubbyResp.Data == nil {
+		return logical.ErrorResponse("wrapping information was nil; wrapping token may be from a previous Vault version"), nil
+	}
+
+	response := cubbyResp.Data["response"]
+	if response == nil {
+		return logical.ErrorResponse("no response found inside the cubbyhole"), nil
+	}
+
+	// TODO: need to figure out the right way forward here; wrapping the response data leads to a double wrap. This needs to either bypass that or decode and copy.
```

Comment:
...Something is odd. I removed that a while ago. Maybe you didn't refresh the page?